File: code-recipes/python/src/recipes/vendor_retail_procurement/shipments/get_shipment_labels_recipe.py
"""
Vendor Shipments API: Get Shipment Labels
==========================================

Retrieves shipping labels using the Vendor Shipments API.
Labels contain carrier information, tracking IDs, and label data (base64 encoded PDF).

Use cases:
- Download shipping labels for printing
- Get carrier codes and tracking information
- Retrieve labels by date range or shipment ID

API operation: getShipmentLabels
"""

import base64
import logging
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.vendor_shipments_v1.vendor_shipment_api import VendorShipmentApi

from src import config
from src.util.recipe import Recipe

logger = logging.getLogger(__name__)


class GetShipmentLabelsRecipe(Recipe):

    def __init__(
        self,
        sp_config: Optional[SPAPIConfig] = None,
        vendor_shipping_api: Optional[VendorShipmentApi] = None,
    ) -> None:
        super().__init__(config=sp_config)
        self._vendor_shipping_api = vendor_shipping_api

    @property
    def vendor_shipping_api(self) -> VendorShipmentApi:
        if self._vendor_shipping_api is None:
            client = SPAPIClient(
                self.config,
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url,
            )
            self._vendor_shipping_api = VendorShipmentApi(client.api_client)
            logger.info("VendorShipmentApi client initialized")
        return self._vendor_shipping_api

    def get_labels_by_date_range(
        self,
        label_created_after: datetime,
        label_created_before: datetime,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Get shipping labels created within a date range."""
        logger.info(
            "Fetching labels: created_after=%s, created_before=%s",
            label_created_after.isoformat(),
            label_created_before.isoformat(),
        )
        response = self.vendor_shipping_api.get_shipment_labels(
            limit=limit,
            sort_order="DESC",
            label_created_after=label_created_after,
            label_created_before=label_created_before,
        )
        labels = self._extract_labels(response)
        logger.info("Found %d labels", len(labels))
        return labels

    def get_labels_by_vendor_shipment_id(self, vendor_shipment_id: str) -> List[Dict[str, Any]]:
        """Get shipping labels by vendor shipment identifier."""
        logger.info("Fetching labels for vendor shipment: %s", vendor_shipment_id)
        response = self.vendor_shipping_api.get_shipment_labels(
            limit=50,
            vendor_shipment_identifier=vendor_shipment_id,
        )
        labels = self._extract_labels(response)
        logger.info("Found %d labels for vendor shipment: %s", len(labels), vendor_shipment_id)
        return labels

    def start(self) -> None:
        now = datetime.now(timezone.utc)
        seven_days_ago = now - timedelta(days=7)

        labels = self.get_labels_by_date_range(
            label_created_after=seven_days_ago,
            label_created_before=now,
        )
        logger.info("Found %d labels", len(labels))

        for label in labels:
            self._log_label_summary(label)

    @staticmethod
    def _extract_labels(response: Any) -> List[Dict[str, Any]]:
        if hasattr(response, "to_dict"):
            response = response.to_dict()
        payload = response.get("payload") or {}
        return payload.get("transport_labels") or []

    @staticmethod
    def _log_label_summary(label: Dict[str, Any]) -> None:
        logger.info("----------------------------------------")
        logger.info("Label Created: %s", label.get("label_create_date_time"))
        shipment_info = label.get("shipment_information") or {}
        if shipment_info:
            logger.info("Buyer Reference: %s", shipment_info.get("buyer_reference_number"))
            vendor_details = shipment_info.get("vendor_details") or {}
            logger.info("Vendor Shipment ID: %s", vendor_details.get("vendor_shipment_identifier"))
            ship_to = shipment_info.get("ship_to_party") or {}
            logger.info("Ship To: %s", ship_to.get("party_id"))

        label_data_list = label.get("label_data") or []
        logger.info("Label Data: %d labels available", len(label_data_list))
        for label_data in label_data_list:
            logger.info(
                "  - Tracking: %s | Carrier: %s | Format: %s",
                label_data.get("tracking_number"),
                label_data.get("carrier_code"),
                label_data.get("label_format"),
            )
            raw_label = label_data.get("label")
            if raw_label:
                pdf_bytes = base64.b64decode(raw_label)
                tracking = label_data.get("tracking_number") or "unknown"
                filename = f"label_{tracking}.pdf"
                file_path = Path(filename).resolve()
                file_path.write_bytes(pdf_bytes)
                logger.info("    Saved: %s (%d bytes)", file_path, len(pdf_bytes))


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    recipe = GetShipmentLabelsRecipe()
    recipe.start()

File: code-recipes/java/src/main/java/vendorretailprocurement/shipments/GetShipmentLabelsRecipe.java
package vendorretailprocurement.shipments;

import com.amazon.SellingPartnerAPIAA.LWAException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.threeten.bp.OffsetDateTime;
import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.vendor.shipments.v1.VendorShippingApi;
import software.amazon.spapi.models.vendor.shipments.v1.GetShipmentLabels;
import software.amazon.spapi.models.vendor.shipments.v1.TransportLabel;
import util.Constants;
import util.Recipe;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.List;

/**
 * Vendor Shipments API Recipe: Get Shipment Labels
 * =================================================
 * 
 * This recipe demonstrates how to retrieve shipping labels using the
 * Vendor Shipments API. Labels contain carrier information, tracking IDs,
 * and the actual label data (base64 encoded).
 * 
 * Use cases:
 * - Download shipping labels for printing
 * - Get carrier codes and tracking information
 * - Retrieve labels by date range or shipment ID
 * 
 * API Operations used:
 * - getShipmentLabels: Returns shipping labels based on filters
 */
public class GetShipmentLabelsRecipe extends Recipe {
    private static final Logger logger = LoggerFactory.getLogger(GetShipmentLabelsRecipe.class);

    private final VendorShippingApi vendorShippingApi = new VendorShippingApi.Builder()
            .lwaAuthorizationCredentials(lwaCredentials)
            .endpoint(Constants.BACKEND_URL)
            .build();

    @Override
    public void start() {
        // Example: Get labels created in the last 7 days
        List<TransportLabel> labels = getLabelsByDateRange(
                OffsetDateTime.now().minusDays(7),
                OffsetDateTime.now()
        );
        logger.info("Found {} labels", labels.size());

        for (TransportLabel label : labels) {
            logLabelSummary(label);
        }
    }

    /**
     * Get shipping labels created within a date range.
     */
    public List<TransportLabel> getLabelsByDateRange(
            OffsetDateTime labelCreatedAfter,
            OffsetDateTime labelCreatedBefore) {
        try {
            GetShipmentLabels response = vendorShippingApi.getShipmentLabels(
                    50L,                  // limit
                    "DESC",               // sortOrder
                    null,                 // nextToken
                    labelCreatedAfter,    // labelCreatedAfter
                    labelCreatedBefore,   // labelCreatedBefore
                    null,                 // buyerReferenceNumber
                    null,                 // vendorShipmentIdentifier
                    null                  // sellerWarehouseCode
            );

            if (response.getPayload() != null && response.getPayload().getTransportLabels() != null) {
                return response.getPayload().getTransportLabels();
            }
            return List.of();
        } catch (ApiException | LWAException e) {
            logger.error("Error getting shipment labels: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to get shipment labels", e);
        }
    }

    /**
     * Get shipping labels by vendor shipment identifier.
     */
    public List<TransportLabel> getLabelsByVendorShipmentId(String vendorShipmentId) {
        try {
            GetShipmentLabels response = vendorShippingApi.getShipmentLabels(
                    50L, null, null, null, null,
                    null,                 // buyerReferenceNumber
                    vendorShipmentId,     // vendorShipmentIdentifier
                    null                  // sellerWarehouseCode
            );

            if (response.getPayload() != null && response.getPayload().getTransportLabels() != null) {
                logger.info("Found {} labels for vendor shipment: {}", 
                        response.getPayload().getTransportLabels().size(), vendorShipmentId);
                return response.getPayload().getTransportLabels();
            }
            logger.warn("No labels found for vendor shipment: {}", vendorShipmentId);
            return List.of();
        } catch (ApiException | LWAException e) {
            logger.error("Error getting labels for shipment {}: {}", vendorShipmentId, e.getMessage(), e);
            throw new RuntimeException("Failed to get labels by vendor shipment ID", e);
        }
    }

    /**
     * Get shipping labels by buyer reference number.
     */
    public List<TransportLabel> getLabelsByBuyerReference(String buyerReferenceNumber) {
        try {
            GetShipmentLabels response = vendorShippingApi.getShipmentLabels(
                    50L, null, null, null, null,
                    buyerReferenceNumber, // buyerReferenceNumber
                    null,                 // vendorShipmentIdentifier
                    null                  // sellerWarehouseCode
            );

            if (response.getPayload() != null && response.getPayload().getTransportLabels() != null) {
                logger.info("Found {} labels for buyer reference: {}", 
                        response.getPayload().getTransportLabels().size(), buyerReferenceNumber);
                return response.getPayload().getTransportLabels();
            }
            logger.warn("No labels found for buyer reference: {}", buyerReferenceNumber);
            return List.of();
        } catch (ApiException | LWAException e) {
            logger.error("Error getting labels for buyer reference {}: {}", buyerReferenceNumber, e.getMessage(), e);
            throw new RuntimeException("Failed to get labels by buyer reference", e);
        }
    }

    private void logLabelSummary(TransportLabel label) {
        logger.info("----------------------------------------");
        logger.info("Label Created: {}", label.getLabelCreateDateTime());
        
        if (label.getShipmentInformation() != null) {
            var shipInfo = label.getShipmentInformation();
            logger.info("Buyer Reference: {}", shipInfo.getBuyerReferenceNumber());
            
            if (shipInfo.getVendorDetails() != null) {
                logger.info("Vendor Shipment ID: {}", shipInfo.getVendorDetails().getVendorShipmentIdentifier());
            }
            
            if (shipInfo.getShipToParty() != null) {
                logger.info("Ship To: {}", shipInfo.getShipToParty().getPartyId());
            }
        }
        
        if (label.getLabelData() != null && !label.getLabelData().isEmpty()) {
            logger.info("Label Data: {} labels available", label.getLabelData().size());
            
            // Demonstrate how to decode and use the label PDF
            for (var labelData : label.getLabelData()) {
                logger.info("  - Seq: {} | Format: {} | Carrier: {} | Tracking: {}",
                        labelData.getLabelSequenceNumber(),
                        labelData.getLabelFormat(),
                        labelData.getCarrierCode(),
                        labelData.getTrackingId());
                
                // The label is base64-encoded PDF - decode it to bytes
                if (labelData.getLabel() != null && !labelData.getLabel().isEmpty()) {
                    byte[] pdfBytes = Base64.getDecoder().decode(labelData.getLabel());
                    
                    // Save the PDF to current directory
                    String filename = "label_" + labelData.getTrackingId() + ".pdf";
                    Path filePath = Path.of(filename).toAbsolutePath();
                    try {
                        Files.write(filePath, pdfBytes);
                        logger.info("    Saved: {} ({} bytes)", filePath, pdfBytes.length);
                    } catch (IOException e) {
                        logger.error("    Failed to save {}: {}", filePath, e.getMessage());
                    }
                }
            }
        }
    }
}
